% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/JAM_LogisticToLinearEffects.R
\name{JAM_LogisticToLinearEffects}
\alias{JAM_LogisticToLinearEffects}
\title{Logistic to linear effect conversion for JAM}
\usage{
JAM_LogisticToLinearEffects(
  log.ors = NULL,
  log.or.ses = NULL,
  snp.genotype.sds = NULL,
  mafs = NULL,
  n = NULL,
  p.cases = NULL
)
}
\arguments{
\item{log.ors}{Vector of SNP summary log-odds ratios}

\item{log.or.ses}{Vector of the standard errors of the SNP log-odds ratios}

\item{snp.genotype.sds}{Vector of SNP genotype standard deviations. Please supply this or mafs. This is
the preferred option since an assumption of Hardy-Weinberg Equilibrium is not required when
reverse-stamdardising the effects during the transform.}

\item{mafs}{Vector of SNP minor allele frequencies. snp.sds if available would be the preferred option.}

\item{n}{Size of dataset in which the log-odds ratios were calculated}

\item{p.cases}{Proportion of cases in the dataset in which the log-odds ratios were calculated}
}
\value{
A vector of effects on the linear scale
}
\description{
Given marginal log-ORs and corresponding MAFs, this function infers the effect estimates that would have
been obtained running one-at-a-time linear regressions of the binary outcome. This conversion is required for
JAM since it is based on a linear regression model.
}
\author{
Paul Newcombe
}
