<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Function reference â€¢ flashfmZero</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">flashfmZero</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/flashfmZero.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="all-functions">All functions <a href="#all-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="AutocorrelationPlot.html">AutocorrelationPlot()</a></code> </p>
        </td>
        <td><p>Autocorrelation plot for model selection. Model inclusion of each predictor is
represented across all iterations as a black and white heatmap.</p></td>
      </tr><tr><td>
          <p><code><a href="CredibleSet.html">CredibleSet()</a></code> </p>
        </td>
        <td><p>Infer credible sets of predictors</p></td>
      </tr><tr><td>
          <p><code><a href="FLASHFMZEROwithJAMd.html">FLASHFMZEROwithJAMd()</a></code> </p>
        </td>
        <td><p>Wrapper for flashfm Multi-Trait Fine-Mapping with JAM (when trait correlation is zero)  - this is the dynamic number of max causal variant version</p></td>
      </tr><tr><td>
          <p><code><a href="FLASHFMwithFINEMAP.html">FLASHFMwithFINEMAP()</a></code> </p>
        </td>
        <td><p>Wrapper to run single-trait fine-mapping with FINEMAP on each trait, followed by flashfm and then constuct SNP groups for each approach and summarises results</p></td>
      </tr><tr><td>
          <p><code><a href="FLASHFMwithJAM.html">FLASHFMwithJAM()</a></code> </p>
        </td>
        <td><p>Wrapper to run single-trait fine-mapping with JAMexpandedCor.multi on each trait, followed by flashfm and then constuct SNP groups for each approach and summarises results</p></td>
      </tr><tr><td>
          <p><code><a href="FLASHFMwithJAMd.html">FLASHFMwithJAMd()</a></code> </p>
        </td>
        <td><p>Wrapper for flashfm Multi-Trait Fine-Mapping with JAM  - this is the dynamic number of max causal variant version</p></td>
      </tr><tr><td>
          <p><code><a href="FLASHFMwithJAMhat.html">FLASHFMwithJAMhat()</a></code> </p>
        </td>
        <td><p>Wrapper to run single-trait fine-mapping with JAMexpandedCor.multi on each trait, followed by flashfm (using fast approximation version) and then constuct SNP groups for each approach and summarises results</p></td>
      </tr><tr><td>
          <p><code><a href="FMconfig.html">FMconfig</a></code> </p>
        </td>
        <td><p>List of FINEMAP config file contents for two simulated traits</p></td>
      </tr><tr><td>
          <p><code><a href="GetBetaBinomMuSd.html">GetBetaBinomMuSd()</a></code> </p>
        </td>
        <td><p>Derives the beta-binomial mean and standard deviation</p></td>
      </tr><tr><td>
          <p><code><a href="GetBetaBinomParams.html">GetBetaBinomParams()</a></code> </p>
        </td>
        <td><p>Derives the beta-binomial parameters for target mean and standard deviation</p></td>
      </tr><tr><td>
          <p><code><a href="JAM.html">JAM()</a></code> </p>
        </td>
        <td><p>JAM (Joint Analysis of Marginal statistics)</p></td>
      </tr><tr><td>
          <p><code><a href="JAMPred.html">JAMPred()</a></code> </p>
        </td>
        <td><p>JAMPred</p></td>
      </tr><tr><td>
          <p><code><a href="JAMPred_ParallelBlockIndices.html">JAMPred_ParallelBlockIndices()</a></code> </p>
        </td>
        <td><p>List of which LD blocks are to be analysed on which CPU core</p></td>
      </tr><tr><td>
          <p><code><a href="JAMPred_SplitIntoPositiveDefiniteBlocks.html">JAMPred_SplitIntoPositiveDefiniteBlocks()</a></code> </p>
        </td>
        <td><p>Splits a list of SNPs into positive definite blocks.</p></td>
      </tr><tr><td>
          <p><code><a href="JAM_LogisticToLinearEffects.html">JAM_LogisticToLinearEffects()</a></code> </p>
        </td>
        <td><p>Logistic to linear effect conversion for JAM</p></td>
      </tr><tr><td>
          <p><code><a href="JAM_PointEstimates.html">JAM_PointEstimates()</a></code> <code><a href="JAM_PointEstimates.html">JAM_PosteriorSampler()</a></code> </p>
        </td>
        <td><p>JAM (Joint Analysis of Marginal statistics) multivariate point estimator.</p></td>
      </tr><tr><td>
          <p><code><a href="JAM_PointEstimates_Xcov.html">JAM_PointEstimates_Xcov()</a></code> </p>
        </td>
        <td><p>internal function for multibeta, modified from JAM_PointEstimates_Package_Simplified of R2BGLiMS package</p></td>
      </tr><tr><td>
          <p><code><a href="JAM_PointEstimates_updated.html">JAM_PointEstimates_updated()</a></code> </p>
        </td>
        <td><p>internal function for multibeta, modified from JAM_PointEstimates_Package_Simplified of R2BGLiMS package (Paul Newcombe)</p></td>
      </tr><tr><td>
          <p><code><a href="JAM_RankCheck.html">JAM_RankCheck()</a></code> </p>
        </td>
        <td><p>JAM rank check.</p></td>
      </tr><tr><td>
          <p><code><a href="JAMdwithGroups.html">JAMdwithGroups()</a></code> </p>
        </td>
        <td><p>Wrapper for JAMdynamic single-trait fine-mapping that also outputs SNP groups; PP and credible sets include SNP group information</p></td>
      </tr><tr><td>
          <p><code><a href="JAMdynamic.html">JAMdynamic()</a></code> </p>
        </td>
        <td><p>Expanded version of JAM (a single-trait fine-mapping approach) that first runs on thinned SNPs and then expands models on tag SNPs
This version starts at a low upper bound for max causal variants and decides on max upper bound based on data</p></td>
      </tr><tr><td>
          <p><code><a href="JAMexpandedCor.multi.html">JAMexpandedCor.multi()</a></code> </p>
        </td>
        <td><p>Expanded version of JAM (a single-trait fine-mapping approach) that first runs on thinned SNPs and then expands models on tag SNPs; this can run independently on multiple traits</p></td>
      </tr><tr><td>
          <p><code><a href="JAMexpandedCor.multi2.html">JAMexpandedCor.multi2()</a></code> </p>
        </td>
        <td><p>Expanded version of JAM (a single-trait fine-mapping approach) that first runs on thinned SNPs and then expands models on tag SNPs; this can run independently on multiple traits
This version is more stable than JAMexpandedCor.multi, but slower, so is run only if JAMexpandedCor.multi fails</p></td>
      </tr><tr><td>
          <p><code><a href="ManhattanPlot.html">ManhattanPlot()</a></code> </p>
        </td>
        <td><p>Manhattan Plot</p></td>
      </tr><tr><td>
          <p><code><a href="ModelSizeBayesFactors.html">ModelSizeBayesFactors()</a></code> </p>
        </td>
        <td><p>Infer credible sets of predictors</p></td>
      </tr><tr><td>
          <p><code><a href="Neff.html">Neff()</a></code> </p>
        </td>
        <td><p>Approximate effective sample size from GWAS of related samples</p></td>
      </tr><tr><td>
          <p><code><a href="NormInvGamPosteriorSample.html">NormInvGamPosteriorSample()</a></code> </p>
        </td>
        <td><p>Generate conjugate posterior sample of coefficients for a particular linear model.</p></td>
      </tr><tr><td>
          <p><code><a href="PPsummarise.html">PPsummarise()</a></code> </p>
        </td>
        <td><p>Summarise PP and MPP results from single-trait fine-mapping and flashfm, by SNP and by SNP group</p></td>
      </tr><tr><td>
          <p><code><a href="PrettyResultsTable.html">PrettyResultsTable()</a></code> </p>
        </td>
        <td><p>A pretty summary results table for reports. NOTE: This function outputs a table formatted with character strings.
A numeric representation of the results are stored in the slot 'posterior.summary.table'.</p></td>
      </tr><tr><td>
          <p><code><a href="R2BGLiMS.html">R2BGLiMS()</a></code> </p>
        </td>
        <td><p>Call BGLiMS from R</p></td>
      </tr><tr><td>
          <p><code><a href="R2BGLiMS_Results-class.html">R2BGLiMS_Results-class</a></code> <code><a href="R2BGLiMS_Results-class.html">R2BGLiMS_Results,R2BGLiMS_Results-class</a></code> </p>
        </td>
        <td><p>The R2BGLiMS_Results class</p></td>
      </tr><tr><td>
          <p><code><a href="R2BGLiMS_Results-methods.html">show(<i>&lt;R2BGLiMS_Results&gt;</i>)</a></code> </p>
        </td>
        <td><p>'show' method for R2BGLiMS-R2BGLiMS_Results-class objects</p></td>
      </tr><tr><td>
          <p><code><a href="SMIM1.N.html">SMIM1.N</a></code> </p>
        </td>
        <td><p>Sample size for the subset of the INTERVAL data that has no missing trait measurements
Original INTERVAL GWAS data are available in <a href="https://doi.org/10.1016/j.cell.2016.10.042" class="external-link">Astle et al. (2016)</a>
and <a href="https://doi.org/10.1038/s41467-023-40679-y" class="external-link">Akbari et al. (2023)</a></p></td>
      </tr><tr><td>
          <p><code><a href="SMIM1.corX.html">SMIM1.corX</a></code> </p>
        </td>
        <td><p>SNP correlation matrix for the subset of the INTERVAL data that has no missing trait measurements
Original INTERVAL GWAS data are available in <a href="https://doi.org/10.1016/j.cell.2016.10.042" class="external-link">Astle et al. (2016)</a>
and <a href="https://doi.org/10.1038/s41467-023-40679-y" class="external-link">Akbari et al. (2023)</a></p></td>
      </tr><tr><td>
          <p><code><a href="SMIM1.gwas.list.interval.fa25.html">SMIM1.gwas.list.interval.fa25</a></code> </p>
        </td>
        <td><p>GWAS data for three latent traits from a subset of the INTERVAL data that has no missing trait measurements
This is a list of three latent trait GWAS data.frames, where each data.frame has columns rsID, beta, EAF, P_value
These latent traits (ML4, ML12, ML14) are all related to red blood cell traits and our analyses suggest that
rs1175550 is the causal variant for all three traits, which is in agreement with UKBB fine-mapping <a href="https://doi.org/10.1016/j.cell.2020.08.008" class="external-link">(Vuckovic et al. 2020)</a> previous studies <a href="https://doi.org/10.1038/ng.2603" class="external-link">(Cvejic et al. 2013)</a>
Original INTERVAL GWAS data are available in <a href="https://doi.org/10.1016/j.cell.2016.10.042" class="external-link">Astle et al. (2016)</a>
and <a href="https://doi.org/10.1038/s41467-023-40679-y" class="external-link">Akbari et al. (2023)</a></p></td>
      </tr><tr><td>
          <p><code><a href="StrongestPairwiseCorrelation.html">StrongestPairwiseCorrelation()</a></code> </p>
        </td>
        <td><p>Strongest Pairwise Correlation</p></td>
      </tr><tr><td>
          <p><code><a href="Sxy.hat.html">Sxy.hat()</a></code> </p>
        </td>
        <td><p>estimates cross-product of each SNP with one trait</p></td>
      </tr><tr><td>
          <p><code><a href="TopModels.html">TopModels()</a></code> </p>
        </td>
        <td><p>Table of top models</p></td>
      </tr><tr><td>
          <p><code><a href="TracePlots.html">TracePlots()</a></code> </p>
        </td>
        <td><p>Parameter posterior trace plots for a R2BGLiMS results object</p></td>
      </tr><tr><td>
          <p><code><a href="Vres.all.html">Vres.all()</a></code> </p>
        </td>
        <td><p>variance of model residuals for trait T1 at all models that have joint effect estimates</p></td>
      </tr><tr><td>
          <p><code><a href="Vres.hat.html">Vres.hat()</a></code> </p>
        </td>
        <td><p>variance of model residuals for trait T1 at model index imod</p></td>
      </tr><tr><td>
          <p><code><a href="Vx.hat.html">Vx.hat()</a></code> </p>
        </td>
        <td><p>Internal function, Vx.hat</p></td>
      </tr><tr><td>
          <p><code><a href="accessors.html">snps()</a></code> <code><a href="accessors.html">tags()</a></code> </p>
        </td>
        <td><p>Accessors for groups objects</p></td>
      </tr><tr><td>
          <p><code><a href="allC12.html">allC12()</a></code> </p>
        </td>
        <td><p>internal function for calcAdjPP for that gives list of covariance matrix of residuals for all trait pairs</p></td>
      </tr><tr><td>
          <p><code><a href="allcredsets.html">allcredsets()</a></code> </p>
        </td>
        <td><p>Construct a credible set for each trait and under each of single and multi-trait fine-mapping</p></td>
      </tr><tr><td>
          <p><code><a href="allcredsetsPP.html">allcredsetsPP()</a></code> </p>
        </td>
        <td><p>Construct a credible set for each trait and under each of single and multi-trait fine-mapping, and provide SNP PP details</p></td>
      </tr><tr><td>
          <p><code><a href="best.models.cpp.html">best.models.cpp()</a></code> </p>
        </td>
        <td><p>Best models from a snpmpd object by cpp or maximum number of models - modification of best.models from GUESSFM by Chris Wallace</p></td>
      </tr><tr><td>
          <p><code><a href="best.snps.html">best.snps()</a></code> </p>
        </td>
        <td><p>Best SNPs</p></td>
      </tr><tr><td>
          <p><code><a href="calc.maxmin.html">calc.maxmin()</a></code> </p>
        </td>
        <td><p>calculate max or min of subset of a matrix</p></td>
      </tr><tr><td>
          <p><code><a href="calcABF.html">calcABF()</a></code> </p>
        </td>
        <td><p>Calculate approximate Bayes' factor (ABF)</p></td>
      </tr><tr><td>
          <p><code><a href="calcAdjPP.html">calcAdjPP()</a></code> </p>
        </td>
        <td><p>Calculates trait-adjusted posterior probabilities for all traits at sharing parameter kappa</p></td>
      </tr><tr><td>
          <p><code><a href="calcCres12.html">calcCres12()</a></code> </p>
        </td>
        <td><p>covariance between residuals of a pair of models for  a trait pair</p></td>
      </tr><tr><td>
          <p><code><a href="calcD12.html">calcD12()</a></code> </p>
        </td>
        <td><p>internal function for calcAdjPP for the 2-trait case</p></td>
      </tr><tr><td>
          <p><code><a href="calcDcon.html">calcDcon()</a></code> </p>
        </td>
        <td><p>internal function for calcAdjPP that gives constant term for delta</p></td>
      </tr><tr><td>
          <p><code><a href="calcQ12.html">calcQ12()</a></code> </p>
        </td>
        <td><p>internal function for calcAdjPP for a pair of traits</p></td>
      </tr><tr><td>
          <p><code><a href="conversion.html">convert()</a></code> </p>
        </td>
        <td><p>Convert from old definitions of groups, tags classes to new</p></td>
      </tr><tr><td>
          <p><code><a href="cor.refdata.fn.html">cor.refdata.fn()</a></code> </p>
        </td>
        <td><p>Thin genotype correlation matrix for JAM input</p></td>
      </tr><tr><td>
          <p><code><a href="cor2cov.html">cor2cov()</a></code> </p>
        </td>
        <td><p>Correlation Matrix to Covariance Matrix Conversion</p></td>
      </tr><tr><td>
          <p><code><a href="credset.html">credset()</a></code> </p>
        </td>
        <td><p>Construct a credible set for a trait</p></td>
      </tr><tr><td>
          <p><code><a href="dot-ApproxPostProbsByModelEnumeration.html">.ApproxPostProbsByModelEnumeration()</a></code> </p>
        </td>
        <td><p>Enumertated approximate posterior probabilities</p></td>
      </tr><tr><td>
          <p><code><a href="dot-BayesFactor.html">.BayesFactor()</a></code> </p>
        </td>
        <td><p>Calculates a Bayes Facotor, give a prior and posterior probabilities.</p></td>
      </tr><tr><td>
          <p><code><a href="dot-BetaBinomialProbabilitySpecificModel.html">.BetaBinomialProbabilitySpecificModel()</a></code> </p>
        </td>
        <td><p>Calculates a Beta-Binomial prior probability for a SPECIFIC model. From Bottolo et al.
This is not the prior on a particular dimension (that would required the binomial co-efficient).</p></td>
      </tr><tr><td>
          <p><code><a href="dot-ConfounderAdjustedResiduals.html">.ConfounderAdjustedResiduals()</a></code> </p>
        </td>
        <td><p>Replaces the outcome variable with residuals from a linear regression on the confounders. I.e. removes effect
of confounders for the conjugate models.</p></td>
      </tr><tr><td>
          <p><code><a href="dot-ModelSpaceProbs.html">.ModelSpaceProbs()</a></code> </p>
        </td>
        <td><p>Calculates prior probababilities for x, and &gt;=x causal variables, when a truncated Poisson prior is used for model space</p></td>
      </tr><tr><td>
          <p><code><a href="dot-ModelSpaceSpecProb.html">.ModelSpaceSpecProb()</a></code> </p>
        </td>
        <td><p>Calculates prior probabability of causality for a particular variable, when a Poisson prior is used for model space</p></td>
      </tr><tr><td>
          <p><code><a href="dot-ReadData.html">.ReadData()</a></code> </p>
        </td>
        <td><p>Read formatted data file</p></td>
      </tr><tr><td>
          <p><code><a href="dot-ReweightSnpEffectsAccordingToBlockCorrelations.html">.ReweightSnpEffectsAccordingToBlockCorrelations()</a></code> </p>
        </td>
        <td><p>This function is used by JAMPred_Step2AdjustmentAndPredictions</p></td>
      </tr><tr><td>
          <p><code><a href="dot-WriteData.html">.WriteData()</a></code> </p>
        </td>
        <td><p>Write Java MCMC format data file</p></td>
      </tr><tr><td>
          <p><code><a href="expand.mod.html">expand.mod()</a></code> </p>
        </td>
        <td><p>internal function for expanding models by tag SNPs in JAMexpanded.multi</p></td>
      </tr><tr><td>
          <p><code><a href="finemap.html">finemap()</a></code> </p>
        </td>
        <td><p>Wrapper to run FINEMAP  (Benner et al. 2016) in R</p></td>
      </tr><tr><td>
          <p><code><a href="flashfm.html">flashfm()</a></code> </p>
        </td>
        <td><p>Marginal PP for models of a set of traits, sharing information between the traits</p></td>
      </tr><tr><td>
          <p><code><a href="flashfm.input.html">flashfm.input()</a></code> </p>
        </td>
        <td><p>Key input for flashfm - constructs snpmod object list and joint effect estimates list for all trait if have external single trait fine-mapping results</p></td>
      </tr><tr><td>
          <p><code><a href="flashfmZero.html">flashfmZero()</a></code> </p>
        </td>
        <td><p>Marginal PP for models of a set of traits, sharing information between the traits, when trait correlation is zero</p></td>
      </tr><tr><td>
          <p><code><a href="flashfmZero.input.html">flashfmZero.input()</a></code> </p>
        </td>
        <td><p>Key input for flashfm - constructs snpmod object list and joint effect estimates list for all trait if have external single trait fine-mapping results, when trait correlation is zero</p></td>
      </tr><tr><td>
          <p><code><a href="groupIDs.fn.html">groupIDs.fn()</a></code> </p>
        </td>
        <td><p>Find SNP group ids for a set of SNPs</p></td>
      </tr><tr><td>
          <p><code><a href="groupmulti.html">groupmulti()</a></code> </p>
        </td>
        <td><p>Group SNPs; adapted from group.multi by Chris Wallace</p></td>
      </tr><tr><td>
          <p><code><a href="groupmultiU.html">groupmultiU()</a></code> </p>
        </td>
        <td><p>Group SNPs; adapted from group.multi by Chris Wallace</p></td>
      </tr><tr><td>
          <p><code><a href="groups-class.html">groups-class</a></code> </p>
        </td>
        <td><p>Group focused class for holding information about sets of SNPs
defined by their mutual LD</p></td>
      </tr><tr><td>
          <p><code><a href="groups-subset.html">tagsof()</a></code> <code><a href="groups-subset.html">taggedby()</a></code> <code><a href="groups-subset.html">`[`(<i>&lt;groups&gt;</i>,<i>&lt;character&gt;</i>,<i>&lt;missing&gt;</i>,<i>&lt;missing&gt;</i>)</a></code> <code><a href="groups-subset.html">`[`(<i>&lt;groups&gt;</i>,<i>&lt;numeric&gt;</i>,<i>&lt;missing&gt;</i>,<i>&lt;missing&gt;</i>)</a></code> <code><a href="groups-subset.html">`[`(<i>&lt;groups&gt;</i>,<i>&lt;logical&gt;</i>,<i>&lt;missing&gt;</i>,<i>&lt;missing&gt;</i>)</a></code> <code><a href="groups-subset.html">`[`(<i>&lt;tags&gt;</i>,<i>&lt;character&gt;</i>,<i>&lt;missing&gt;</i>,<i>&lt;missing&gt;</i>)</a></code> <code><a href="groups-subset.html">`[[`(<i>&lt;groups&gt;</i>,<i>&lt;numeric&gt;</i>)</a></code> <code><a href="groups-subset.html">`[[`(<i>&lt;groups&gt;</i>,<i>&lt;logical&gt;</i>)</a></code> <code><a href="groups-subset.html">`[[`(<i>&lt;groups&gt;</i>,<i>&lt;character&gt;</i>)</a></code> </p>
        </td>
        <td><p><code>tagsof</code> shows tags for a named character vector of SNPs</p></td>
      </tr><tr><td>
          <p><code><a href="logsum.html">logsum()</a></code> </p>
        </td>
        <td><p>logsum</p></td>
      </tr><tr><td>
          <p><code><a href="makeNlist.html">makeNlist()</a></code> </p>
        </td>
        <td><p>Sample size information needed for flashfm</p></td>
      </tr><tr><td>
          <p><code><a href="makeNlist.rel.html">makeNlist.rel()</a></code> </p>
        </td>
        <td><p>Sample size information needed for flashfm, when samples are related and have effective sample sizes for each trait</p></td>
      </tr><tr><td>
          <p><code><a href="makeSNPgroups.html">makeSNPgroups()</a></code> </p>
        </td>
        <td><p>Make SNP groups using fine-mapping information from all of the traits</p></td>
      </tr><tr><td>
          <p><code><a href="makeSNPgroups2.html">makeSNPgroups2()</a></code> </p>
        </td>
        <td><p>Make two sets of SNP groups using fine-mapping information from all of the traits using two sets of results and maps the names between them</p></td>
      </tr><tr><td>
          <p><code><a href="makeSNPgroups2U.html">makeSNPgroups2U()</a></code> </p>
        </td>
        <td><p>Make two sets of SNP groups using fine-mapping information from all of the traits using two sets of results and maps the names between them</p></td>
      </tr><tr><td>
          <p><code><a href="makeSNPgroupsU.html">makeSNPgroupsU()</a></code> </p>
        </td>
        <td><p>Make SNP groups using fine-mapping information from all of the traits</p></td>
      </tr><tr><td>
          <p><code><a href="makemod.html">makemod()</a></code> </p>
        </td>
        <td><p>internal function makemod</p></td>
      </tr><tr><td>
          <p><code><a href="marg.snps.html">marg.snps()</a></code> </p>
        </td>
        <td><p>internal function marg.snps</p></td>
      </tr><tr><td>
          <p><code><a href="marg.snps.vecs.html">marg.snps.vecs()</a></code> </p>
        </td>
        <td><p>internal function marg.snps.vec</p></td>
      </tr><tr><td>
          <p><code><a href="marginalpp.html">marginalpp()</a></code> </p>
        </td>
        <td><p>Marginal PP for models sharing information between traits</p></td>
      </tr><tr><td>
          <p><code><a href="marginalpp0.html">marginalpp0()</a></code> </p>
        </td>
        <td><p>Marginal PP for models sharing information between traits, when trait correlation is zero</p></td>
      </tr><tr><td>
          <p><code><a href="mod.fn.html">mod.fn()</a></code> </p>
        </td>
        <td><p>internal processing function for JAMexpanded.multi</p></td>
      </tr><tr><td>
          <p><code><a href="multiJAMd.html">multiJAMd()</a></code> </p>
        </td>
        <td><p>Wrapper for JAMdynamic single-trait fine-mapping applied to a set of traits, and output SNP groups</p></td>
      </tr><tr><td>
          <p><code><a href="multiJAMdCS.html">multiJAMdCS()</a></code> </p>
        </td>
        <td><p>Return credible sets from single-trait fine-mapping of several traits</p></td>
      </tr><tr><td>
          <p><code><a href="multibeta.html">multibeta()</a></code> </p>
        </td>
        <td><p>Using summary statistics, calculates joint effect estimates</p></td>
      </tr><tr><td>
          <p><code><a href="overlap.html">overlap()</a></code> </p>
        </td>
        <td><p>overlap</p></td>
      </tr><tr><td>
          <p><code><a href="ppnsnp-class.html">ppnsnp-class</a></code> </p>
        </td>
        <td><p>Class to hold results of pp.nsnp</p></td>
      </tr><tr><td>
          <p><code><a href="show-methods.html">show(<i>&lt;snpmod&gt;</i>)</a></code> <code><a href="show-methods.html">show(<i>&lt;snppicker&gt;</i>)</a></code> <code><a href="show-methods.html">show(<i>&lt;ppnsnp&gt;</i>)</a></code> <code><a href="show-methods.html">show(<i>&lt;tags&gt;</i>)</a></code> <code><a href="show-methods.html">show(<i>&lt;groups&gt;</i>)</a></code> </p>
        </td>
        <td><p>Show</p></td>
      </tr><tr><td>
          <p><code><a href="snpin.html">snpin()</a></code> </p>
        </td>
        <td><p>Check whether a snp is in a snppicker, groups or tags object</p></td>
      </tr><tr><td>
          <p><code><a href="snpmod-class.html">snpmod-class</a></code> </p>
        </td>
        <td><p>Class to hold data relating to multiple models fitted to SNP data</p></td>
      </tr><tr><td>
          <p><code><a href="snppicker-class.html">snppicker-class</a></code> </p>
        </td>
        <td><p>Class to hold results of snp.picker algorithm</p></td>
      </tr><tr><td>
          <p><code><a href="summary.html">summary(<i>&lt;snppicker&gt;</i>)</a></code> <code><a href="summary.html">summary(<i>&lt;groups&gt;</i>)</a></code> </p>
        </td>
        <td><p>Summaries</p></td>
      </tr><tr><td>
          <p><code><a href="summaryStats.html">summaryStats()</a></code> </p>
        </td>
        <td><p>Summary statistics needed for flashfm input</p></td>
      </tr><tr><td>
          <p><code><a href="tagexpand.mod.html">tagexpand.mod()</a></code> </p>
        </td>
        <td><p>internal function for expanding models by tag SNPs in JAMexpandedCor.multi</p></td>
      </tr><tr><td>
          <p><code><a href="tags-class.html">tags-class</a></code> </p>
        </td>
        <td><p>Tags focused class for holding information about sets of SNPs
defined by their mutual LD</p></td>
      </tr><tr><td>
          <p><code><a href="union.html">union()</a></code> </p>
        </td>
        <td><p>Create a union of groups, snppicker or tags objects</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Jenn Asimit, Feng Zhou.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

